pkg/importer/data_processor.go

```diff
@@ -0,0 +1,211 @@
+	Transfer(path string) (ProcessingPhase, error)
+	// Process() is called to do any special processing before giving the url to the data back to the processor
+	Process() (ProcessingPhase, error)
+	// Geturl returns the url that the data processor can use when converting the data.
+	GetURL() *url.URL
+	// Close closes any readers or other open resources.
+	Close() error
+}
+
+// DataProcessor holds the fields needed to process data from a data provider.
+type DataProcessor struct {
+	// currentPhase is the phase the processing is in currently.
+	currentPhase ProcessingPhase
+	// provider provides the data for processing.
+	provider DataProviderInterface
+	// destination file. will be DataDir/disk.img if file system, or a block device (if a block device, then DataDir will not exist).
+	dest string
+	// dataDir path to target directory if it contains a file system.
+	dataDir string
+	// scratchDataDir path to the scratch space.
+	scratchDataDir string
+	// requestImageSize is the size we want the resulting image to be.
+	requestImageSize string
+	// available space is the available space before downloading the image
+	availableSpace int64
+}
+
+// NewDataProcessor create a new instance of a data processor using the passed in data provider.
+func NewDataProcessor(dataProvider DataProviderInterface, dest, dataDir, scratchDataDir, requestImageSize string) *DataProcessor {
+	dp := &DataProcessor{
+		currentPhase:     Info,
+		provider:         dataProvider,
+		dest:             dest,
+		dataDir:          dataDir,
+		scratchDataDir:   scratchDataDir,
+		requestImageSize: requestImageSize,
+	}
+	// Calculate available space before doing anything.
+	dp.availableSpace = dp.calculateTargetSize()
+	return dp
+}
+
+// ProcessData is the main processing loop.
+func (dp *DataProcessor) ProcessData() error {
+	var err error
+	defer dp.provider.Close()
+	if util.GetAvailableSpace(dp.scratchDataDir) > int64(0) {
+		defer CleanDir(dp.scratchDataDir)
+	}
+	for dp.currentPhase != Complete {
+		switch dp.currentPhase {
+		case Info:
+			dp.currentPhase, err = dp.provider.Info()
+		case Transfer:
+			dp.currentPhase, err = dp.provider.Transfer(dp.scratchDataDir)
+		case TransferTarget:
+			dp.currentPhase, err = dp.provider.Transfer(dp.dataDir)
+		case Process:
+			dp.currentPhase, err = dp.provider.Process()
+		case Convert:
+			dp.currentPhase, err = dp.convert(dp.provider.GetURL())
+		case Resize:
+			dp.currentPhase, err = dp.resize()
+		default:
+			return errors.Errorf("Unknown processing phase %s", dp.currentPhase)
+		}
+		if err != nil {
+			klog.Errorf("%v", err)
+			return err
+		}
+		klog.V(1).Infof("New phase: %s\n", dp.currentPhase)
+	}
+	return err
+}
+
+// convert is called when convert the image from the url to a RAW disk image. Source formats include RAW/QCOW2 (Raw to raw conversion is a copy)
+func (dp *DataProcessor) convert(url *url.URL) (ProcessingPhase, error) {
```

Comment:
I believe we have had this discussion before. Yes technically raw doesn't need to be converted, because its raw already. But the added complexity in code to detect raw and then go into a special code path just for that is simply not worth it. This keeps the logic of the code much simpler. Unless there are significant benefits to a special code path I don't want to make the code more complex.